iwyu.cc

```diff
@@ -2006,28 +2006,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // compiler a fighting chance of generating correct code.
   bool VisitCastExpr(clang::CastExpr* expr) {
     if (CanIgnoreCurrentASTNode())  return true;
-    const Type* const from_type = GetTypeOf(expr->getSubExprAsWritten());
-    const Type* const to_type = GetTypeOf(expr);
-    const Type* const deref_from_type = RemovePointersAndReferences(from_type);
-    const Type* const deref_to_type = RemovePointersAndReferences(to_type);
-
-    // For all those casts that don't result in function calls
-    // (everything except a user-defined cast or a constructor cast),
-    // we only care about the need for full types when casting either
-    // a pointer to a pointer, or any type to a reference.
-    // Unfortunately, when casting to a reference, clang seems to
-    // strip the reference off of to_type, so we need a separate
-    // function call to tell.
-    if (expr->getCastKind() != clang::CK_UserDefinedConversion &&
-        expr->getCastKind() != clang::CK_ConstructorConversion) {
-      if (!((from_type->hasPointerRepresentation() &&   // pointer or reference
-             to_type->hasPointerRepresentation()) ||
-            IsCastToReferenceType(expr)))
-        return true;    // we only care about ptr-to-ptr casts for this check
```

Comment:
Is this no longer true? Or covered by the switch?

Reply:
I modified the switch/case to ignore instead of warn for non-pointer-to-pointer cast kinds. That seems to satisfy the test suite, so I optimistically assumed that was a valid way to simplify (otherwise I would have had to factor the converted-to-type into this already-hairy predicate).

Reply:
I see, that makes sense. So, any unsupported cast kind will not enter the switch in the first place, and get simply ignored, right?

Reply:
In the old implementation that was the case. In the new implementation, all cast kinds enter the switch, but most of them just exit without action.

Reply:
Also, it was necessary/simpler to remove this because a cast from `Convertible` to `Base*` is *not* a pointer-to-pointer cast, but will need to be handled as one.